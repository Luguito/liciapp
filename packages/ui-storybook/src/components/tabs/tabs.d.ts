import PropTypes from 'prop-types';
import { LoanProposal } from '@addi-domains/originations';
import { TabsPropTypes, TabsDefaultValues } from './tabs.prop-types';

export interface TabProps {
    proposal: LoanProposal;
    tabKey: string | number;
    headComponent: (isSelected: boolean) => React.ReactNode;
}

export type TabsProps =  PropTypes.InferProps<typeof TabsPropTypes, typeof TabsDefaultValues>;
