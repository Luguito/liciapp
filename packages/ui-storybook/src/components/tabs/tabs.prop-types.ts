import PropTypes from 'prop-types';

export const TabsPropTypes = {
    /** Callback, return tab selected */
    onChange: PropTypes.func,
};

export const TabsDefaultValues = {};
