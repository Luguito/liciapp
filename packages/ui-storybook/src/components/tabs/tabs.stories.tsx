import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import { Tabs, Tab } from './tabs';
import { TabsStory } from './tabs.styled';

storiesOf('Visualization', module)
    .addDecorator(withKnobs)
    .addDecorator(withInfo)
    .add('Tab Basic', () => {
        const callback = action('Tab selected => ');

        const tabs: any= [{
                key: 1,
                headComponent: () => 1,
                contentComponent: <div>'content 1'</div>
            },{
                key: 2,
                headComponent: () => 2,
                contentComponent: <div>'content 2'</div>
            },{
                key: 3,
                headComponent: () => 3,
                contentComponent: <div>'content 3'</div>
            },
        ];

        return (
            <TabsStory>
                <Tabs onChange={callback} defaultSelected={2}>
                    {tabs.map((tab: any) =>
                        <Tab proposal={tab.proposal} key={tab.key} tabKey={tab.key} headComponent={tab.headComponent}>
                            {tab.contentComponent}
                        </Tab>)
                    }
                </Tabs>
            </TabsStory>
        );
    }, {
        info: 'Tabs basic component'
    });
