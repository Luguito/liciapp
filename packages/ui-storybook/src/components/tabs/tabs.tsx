import React, { useState, useEffect, FunctionComponent, ReactElement, memo } from 'react';
import { TabsProps, TabProps } from './tabs.d';
import {
    TabHeader,
    LineSelector,
    TabsContainer,
    MenuContainer,
    ContentContainer
} from './tabs.styled';
import { TabsDefaultValues, TabsPropTypes } from './tabs.prop-types';

export const Tab: FunctionComponent<TabProps> = () => <div></div>;

const TabsComponent: FunctionComponent<TabsProps> = ({ onChange, children, defaultSelected }) => {
    const tabs = React.Children.toArray(children);
    const [selectedKey, setSelectedKey] = useState<number>(0);

    useEffect(() => {
        defaultSelected && setSelectedKey(defaultSelected);
    }, [defaultSelected]);

    // Call onChange
    useEffect(() => {
        tabs.length
        && onChange
        && onChange(tabs[selectedKey].props);
    }, [selectedKey]);

    return (
        <TabsContainer>
            <MenuContainer>
                {!!tabs.length && tabs.map((tab: ReactElement, indexKey: number) => {
                    const isSelected: boolean = indexKey === selectedKey;

                    return (
                        <TabHeader
                            key={`tab_${indexKey}`}
                            id={`tab_${indexKey}`}
                            isSelected={indexKey === selectedKey}
                            onClick={() => setSelectedKey(indexKey)}
                        >
                            {tab.props.headComponent(isSelected)}
                            <LineSelector isSelected={isSelected} index={selectedKey}/>

                        </TabHeader>
                        );
                    })
                }
            </MenuContainer>
            <ContentContainer index={selectedKey}>
                {!!tabs.length && tabs[selectedKey].props.children}
            </ContentContainer>
        </TabsContainer>
    );
};

TabsComponent.propTypes = TabsPropTypes;
TabsComponent.defaultProps = TabsDefaultValues;

export const Tabs = memo(TabsComponent);
