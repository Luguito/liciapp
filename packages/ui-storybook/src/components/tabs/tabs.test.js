import React from 'react';
import { Tabs } from './tabs';
import { shallow } from 'enzyme';

const mocktTabs = [{
        key: '1',
        headComponent: () => <div>1</div>,
        contentComponent: <div>content 1</div>
    },{
        key: '2',
        headComponent: () => <div>2</div>,
        contentComponent: <div>content 2</div>
    },{
        key: '3',
        headComponent: () => <div>3</div>,
        contentComponent: <div>content 3</div>
    },
];

describe('Basic', () => {
    it('Should be valid snapshot', () => {
        const comp = shallow(<Tabs />)
        expect(comp).toMatchSnapshot()
    });

    it ('Show correct tabs', () => {
        const tabs = shallow(<Tabs tabs={mocktTabs} onChange={() => {}}/>)

        expect(tabs.find('#tab_1')).toHaveLength(1);
        expect(tabs.find('#tab_2')).toHaveLength(1);
        expect(tabs.find('#tab_3')).toHaveLength(1);
    });

    it ('Show correct tab content', () => {
        const tabs = shallow(<Tabs tabs={mocktTabs} onChange={() => {}}/>)
        tabs.find('#tab_2').simulate('click');
        expect(tabs.text()).toMatch('content 2');
    });
});
