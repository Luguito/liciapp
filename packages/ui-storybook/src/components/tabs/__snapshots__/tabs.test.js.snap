// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Should be valid snapshot 1`] = `
<styled.div>
  <styled.div>
    0
  </styled.div>
  <styled.div>
    0
  </styled.div>
</styled.div>
`;
